crates/env/src/engine/e2e/xts.rs

```diff
@@ -0,0 +1,360 @@
+    ///
+    /// Returns when the transaction is included in a block. The return value
+    /// contains all events that are associated with this transaction.
+    #[allow(clippy::too_many_arguments)]
+    pub async fn instantiate_with_code(
+        &self,
+        value: E::Balance,
+        gas_limit: crate::types::Gas,
+        storage_deposit_limit: Option<E::Balance>,
+        code: Vec<u8>,
+        data: Vec<u8>,
+        salt: Vec<u8>,
+        signer: &Signer<C>,
+    ) -> TxEvents<C> {
+        let call = subxt::tx::StaticTxPayload::new(
+            "Contracts",
+            "instantiate_with_code",
+            InstantiateWithCode::<E::Balance> {
+                value,
+                gas_limit,
+                storage_deposit_limit,
+                code,
+                data,
+                salt,
+            },
+            Default::default(),
+        )
+        .unvalidated();
+
+        self
+            .client
+            .tx()
+            .sign_and_submit_then_watch_default(&call, signer)
+            .await
+            .map(|tx_progress| {
+                log_info(&format!("signed and submitted tx with hash {:?}", tx_progress.extrinsic_hash()));
+                tx_progress
+            })
+            .unwrap_or_else(|err| {
+                panic!(
+                    "error on call `sign_and_submit_then_watch_default`: {:?}",
+                    err
+                );
+            })
+            // TODO(#xxx) It should be configurable to use `.wait_for_finalized_success` instead.
+            .wait_for_in_block()
+            .await
+            .unwrap_or_else(|err| {
+                panic!(
+                    "error on call `wait_for_in_block`: {:?}",
+                    err
+                );
+            })
+            .fetch_events()
+            .await
+            .unwrap_or_else(|err| {
+                panic!(
+                    "error on call `fetch_events`: {:?}",
+                    err
+                );
+            })
+    }
+
+    /// Dry runs a call of the contract at `contract` with the given parameters.
+    pub async fn call_dry_run(
+        &self,
+        contract: C::AccountId,
+        value: E::Balance,
+        storage_deposit_limit: Option<E::Balance>,
+        data: Vec<u8>,
+    ) -> ContractDryCallResult<E> {
+        let call_request = RpcCallRequest::<C> {
+            origin: contract.clone(),
+            dest: contract,
+            value: value.try_into().expect("unable to convert `value`"),
+            gas_limit: DRY_RUN_GAS_LIMIT.into(),
+            storage_deposit_limit: storage_deposit_limit.map(|l| {
+                l.try_into()
+                    .expect("unable to convert `storage_deposit_limit`")
+            }),
+            input_data: Bytes(data),
+        };
+        let params = rpc_params![call_request];
+        self.ws_client
+            .request("contracts_call", params)
```

Comment:
The dry run should use the `state_call` rpc, since I believe these RPCs will be deprecated (cc @athei). See https://github.com/paritytech/cargo-contract/pull/701

Reply:
More involved than I thought (multiple repos), I've added a ToDo comment, will create a follow-up issue.

Reply:
Why is it involved? You add new code. You can use `state_call` as well here.

Reply:
Missing `Encode` derive in `smart-bench` and `submit`.

Reply:
On which types? AFAIK there is no support for state call in subxt. You need to define them in this crate anyways (as you have done for the RPC arguments).

Reply:
Migrated it to `state_call` now.

Reply:
Thanks for not adding new users to the deprecated API.